File: Project_frontend/server/chatbot.py
import os
import json
import re
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import FAISS
from langchain.schema import Document
import openai

class DocumentChatbot:
    def __init__(self, api_key, model_name="shisa-ai/shisa-v2-llama3.3-70b:free"):
        # Configure OpenAI for OpenRouter
        openai.api_key = api_key
        openai.api_base = "https://openrouter.ai/api/v1"
        self.model_name = model_name
        
        # Initialize embeddings
        self.embedder = HuggingFaceEmbeddings(model_name="all-MiniLM-L6-v2")
        self.db = None
        
    def load_and_process_documents(self, text_file_path, tables_dir):
        """Load text and table documents, create embeddings"""
        # Load text document
        with open(text_file_path, "r", encoding="utf-8") as f:
            doc_text = f.read()

        splitter = RecursiveCharacterTextSplitter(chunk_size=800, chunk_overlap=100)
        text_chunks = splitter.split_text(doc_text)
        text_docs = [{"type": "text", "content": chunk, "source": "ocr_text"} for chunk in text_chunks]
        
        # Load table documents
        table_docs = []
        if os.path.exists(tables_dir):
            for fname in os.listdir(tables_dir):
                if fname.endswith('.json'):
                    with open(os.path.join(tables_dir, fname), "r", encoding="utf-8") as f:
                        rows = json.load(f)
                        table_md = self.table_to_md(rows)
                        table_docs.append({
                            "type": "table",
                            "content": table_md,
                            "source": fname
                        })
        
        # Create FAISS database
        all_docs = []
        for item in text_docs + table_docs:
            all_docs.append(Document(
                page_content=item["content"],
                metadata={"source": item["source"], "type": item["type"]}
            ))

        self.db = FAISS.from_documents(all_docs, self.embedder)
        print(f"[✓] Loaded {len(text_docs)} text chunks and {len(table_docs)} table chunks")
        
    def table_to_md(self, rows):
        """Convert table rows to markdown format"""
        md = []
        for i, row in enumerate(rows):
            line = "| " + " | ".join(str(cell) for cell in row) + " |"
            md.append(line)
            if i == 0:
                md.append("| " + " | ".join(["---"] * len(row)) + " |")
        return "\n".join(md)
    
    def ask(self, query, k=5):
        """Ask a question and get an answer"""
        if not self.db:
            return "Please load documents first using load_and_process_documents()"
            
        # Get relevant documents
        rel_docs = self.db.similarity_search(query, k=k)
        context = "\n\n---\n\n".join([doc.page_content for doc in rel_docs])
        
        prompt = f"""You are an assistant answering questions using the provided documents only.
If the answer is not in the documents, say "I don't know based on the provided documents."

### Query:
{query}

### Documents:
{context}
"""
        
        try:
            response = openai.ChatCompletion.create(
                model=self.model_name,
                messages=[
                    {"role": "system", "content": "You are a helpful assistant."},
                    {"role": "user", "content": prompt}
                ]
            )
            return response['choices'][0]['message']['content']
        except Exception as e:
            return f"Error generating response: {str(e)}"

# Example usage
if __name__ == "__main__":
    # Initialize chatbot
    chatbot = DocumentChatbot(api_key="your-openrouter-api-key")
    
    # Load documents
    chatbot.load_and_process_documents(
        text_file_path="ocr_text_with_tables.txt",
        tables_dir="Altered_tables"
    )
    
    # Test query
    response = chatbot.ask("What is the HRA rate for X class cities?")
    print(response)

File: Project_frontend/server/chatbot.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import axios from 'axios';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

class DocumentChatbot {
  constructor() {
    this.apiKey = process.env.OPENAI_API_KEY;
    this.apiBase = process.env.OPENAI_API_BASE || 'https://openrouter.ai/api/v1';
    this.modelName = 'shisa-ai/shisa-v2-llama3.3-70b:free';
    
    // File paths from environment or defaults
    this.ocrTextPath = process.env.OCR_TEXT_FILE_PATH || path.join(__dirname, '../data/ocr_text_with_tables.txt');
    this.tablesDir = process.env.TABLES_DIR_PATH || path.join(__dirname, '../data/Altered_tables');
    
    this.documents = [];
    this.isLoaded = false;
    
    this.loadDocuments();
  }

  loadDocuments() {
    try {
      // Load main text document
      if (fs.existsSync(this.ocrTextPath)) {
        const textContent = fs.readFileSync(this.ocrTextPath, 'utf-8');
        this.documents.push({
          type: 'text',
          content: textContent,
          source: 'ocr_text_with_tables'
        });
        console.log('[✓] Loaded OCR text document');
      } else {
        console.log('[!] OCR text file not found at:', this.ocrTextPath);
      }

      // Load table documents
      if (fs.existsSync(this.tablesDir)) {
        const tableFiles = fs.readdirSync(this.tablesDir).filter(file => file.endsWith('.json'));
        
        tableFiles.forEach(filename => {
          const filePath = path.join(this.tablesDir, filename);
          const tableData = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
          const tableMarkdown = this.tableToMarkdown(tableData);
          
          this.documents.push({
            type: 'table',
            content: tableMarkdown,
            source: filename
          });
        });
        
        console.log(`[✓] Loaded ${tableFiles.length} table documents`);
      } else {
        console.log('[!] Tables directory not found at:', this.tablesDir);
      }

      this.isLoaded = this.documents.length > 0;
      console.log(`[✓] Total documents loaded: ${this.documents.length}`);
      
    } catch (error) {
      console.error('[ERROR] Failed to load documents:', error);
      this.isLoaded = false;
    }
  }

  tableToMarkdown(rows) {
    if (!rows || rows.length === 0) return '';
    
    const markdown = [];
    rows.forEach((row, index) => {
      const line = '| ' + row.map(cell => String(cell || '')).join(' | ') + ' |';
      markdown.push(line);
      
      // Add header separator after first row
      if (index === 0) {
        const separator = '| ' + row.map(() => '---').join(' | ') + ' |';
        markdown.push(separator);
      }
    });
    
    return markdown.join('\n');
  }

  // Simple keyword-based document search (fallback for when embeddings aren't available)
  searchDocuments(query, maxResults = 5) {
    const queryLower = query.toLowerCase();
    const results = [];

    this.documents.forEach(doc => {
      const contentLower = doc.content.toLowerCase();
      let score = 0;
      
      // Simple scoring based on keyword matches
      const queryWords = queryLower.split(/\s+/);
      queryWords.forEach(word => {
        if (word.length > 2) { // Skip very short words
          const matches = (contentLower.match(new RegExp(word, 'g')) || []).length;
          score += matches;
        }
      });

      if (score > 0) {
        results.push({
          document: doc,
          score: score
        });
      }
    });

    // Sort by score and return top results
    return results
      .sort((a, b) => b.score - a.score)
      .slice(0, maxResults)
      .map(result => result.document);
  }

  async ask(query) {
    if (!this.isLoaded) {
      return "Sorry, the document database is not loaded. Please check if the data files are properly placed.";
    }

    if (!this.apiKey) {
      return "API key not configured. Please set your OpenRouter API key in the .env file.";
    }

    try {
      // Get relevant documents
      const relevantDocs = this.searchDocuments(query, 5);
      
      if (relevantDocs.length === 0) {
        return "I couldn't find relevant information in the documents to answer your question.";
      }

      // Create context from relevant documents
      const context = relevantDocs
        .map(doc => `Source: ${doc.source}\n${doc.content}`)
        .join('\n\n---\n\n');

      const prompt = `You are an assistant answering questions using the provided documents only.
If the answer is not in the documents, say "I don't know based on the provided documents."

### Query:
${query}

### Documents:
${context}`;

      // Make API call to OpenRouter
      const response = await axios.post(
        `${this.apiBase}/chat/completions`,
        {
          model: this.modelName,
          messages: [
            { role: 'system', content: 'You are a helpful assistant.' },
            { role: 'user', content: prompt }
          ],
          temperature: 0,
          max_tokens: 512
        },
        {
          headers: {
            'Authorization': `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json'
          }
        }
      );

      return response.data.choices[0].message.content;

    } catch (error) {
      console.error('Error generating response:', error);
      
      if (error.response?.status === 401) {
        return "Authentication failed. Please check your API key.";
      } else if (error.response?.status === 429) {
        return "Rate limit exceeded. Please try again later.";
      } else {
        return "Sorry, I encountered an error processing your request. Please try again.";
      }
    }
  }

  // Reload documents (useful for development)
  reload() {
    this.documents = [];
    this.isLoaded = false;
    this.loadDocuments();
  }
}

export default DocumentChatbot;